File: 07-dynamic-programming/examples.py
"""
Классические задачи ДП.
Запуск: python examples.py
"""

from functools import lru_cache


# ── Фибоначчи: 3 версии ──────────────────────────────────────────────
def fib_naive(n: int) -> int:
    """Экспонента, для иллюстрации."""
    if n < 2:
        return n
    return fib_naive(n - 1) + fib_naive(n - 2)


@lru_cache(maxsize=None)
def fib_memo(n: int) -> int:
    if n < 2:
        return n
    return fib_memo(n - 1) + fib_memo(n - 2)


def fib_iter(n: int) -> int:
    if n < 2:
        return n
    a, b = 0, 1
    for _ in range(n - 1):
        a, b = b, a + b
    return b


# ── Размен монет: min(c) на сумму S ──────────────────────────────────
def coin_change(coins: list[int], amount: int) -> int:
    INF = float("inf")
    dp = [0] + [INF] * amount
    for s in range(1, amount + 1):
        for c in coins:
            if c <= s and dp[s - c] + 1 < dp[s]:
                dp[s] = dp[s - c] + 1
    return -1 if dp[amount] == INF else dp[amount]


# ── 0/1 рюкзак ───────────────────────────────────────────────────────
def knapsack_01(weights: list[int], values: list[int], capacity: int) -> int:
    """O(N * W), память сжата до O(W)."""
    n = len(weights)
    dp = [0] * (capacity + 1)
    for i in range(n):
        # справа налево, чтобы не переиспользовать предмет
        for w in range(capacity, weights[i] - 1, -1):
            dp[w] = max(dp[w], dp[w - weights[i]] + values[i])
    return dp[capacity]


# ── LCS: длина общей подпоследовательности ───────────────────────────
def lcs(a: str, b: str) -> int:
    n, m = len(a), len(b)
    dp = [[0] * (m + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if a[i - 1] == b[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
    return dp[n][m]


# ── Левенштейн: расстояние редактирования ────────────────────────────
def edit_distance(a: str, b: str) -> int:
    n, m = len(a), len(b)
    dp = [[0] * (m + 1) for _ in range(n + 1)]
    for i in range(n + 1):
        dp[i][0] = i
    for j in range(m + 1):
        dp[0][j] = j
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if a[i - 1] == b[j - 1]:
                dp[i][j] = dp[i - 1][j - 1]
            else:
                dp[i][j] = 1 + min(
                    dp[i - 1][j],      # удалить
                    dp[i][j - 1],      # вставить
                    dp[i - 1][j - 1],  # заменить
                )
    return dp[n][m]


# ── LIS: длина наибольшей возрастающей подпоследовательности ─────────
def lis(arr: list[int]) -> int:
    """O(n log n) через терпеливую раскладку."""
    from bisect import bisect_left

    tails: list[int] = []
    for x in arr:
        i = bisect_left(tails, x)
        if i == len(tails):
            tails.append(x)
        else:
            tails[i] = x
    return len(tails)


if __name__ == "__main__":
    print("fib(20) iter:", fib_iter(20))
    print("fib(30) memo:", fib_memo(30))
    print("coin_change [1,3,4], 6:", coin_change([1, 3, 4], 6))
    print("knapsack:", knapsack_01([2, 3, 4, 5], [3, 4, 5, 6], capacity=5))
    print("lcs ABCBDAB / BDCABA:", lcs("ABCBDAB", "BDCABA"))
    print("edit_distance kitten / sitting:", edit_distance("kitten", "sitting"))
    print("lis [10,9,2,5,3,7,101,18]:", lis([10, 9, 2, 5, 3, 7, 101, 18]))

File: 07-dynamic-programming/examples.go
// Классические задачи ДП на Go.
// Запуск: go run examples.go
package main

import (
	"fmt"
	"sort"
)

// ── Фибоначчи итеративный ──────────────────────────────────────────
func Fib(n int) int {
	if n < 2 {
		return n
	}
	a, b := 0, 1
	for i := 1; i < n; i++ {
		a, b = b, a+b
	}
	return b
}

// ── Размен монет: -1 если невозможно ───────────────────────────────
func CoinChange(coins []int, amount int) int {
	const INF = 1 << 30
	dp := make([]int, amount+1)
	for i := 1; i <= amount; i++ {
		dp[i] = INF
	}
	for s := 1; s <= amount; s++ {
		for _, c := range coins {
			if c <= s && dp[s-c]+1 < dp[s] {
				dp[s] = dp[s-c] + 1
			}
		}
	}
	if dp[amount] == INF {
		return -1
	}
	return dp[amount]
}

// ── 0/1 рюкзак ──────────────────────────────────────────────────────
func Knapsack01(weights, values []int, capacity int) int {
	dp := make([]int, capacity+1)
	for i := 0; i < len(weights); i++ {
		for w := capacity; w >= weights[i]; w-- {
			if dp[w-weights[i]]+values[i] > dp[w] {
				dp[w] = dp[w-weights[i]] + values[i]
			}
		}
	}
	return dp[capacity]
}

// ── LCS ────────────────────────────────────────────────────────────
func LCS(a, b string) int {
	n, m := len(a), len(b)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, m+1)
	}
	for i := 1; i <= n; i++ {
		for j := 1; j <= m; j++ {
			if a[i-1] == b[j-1] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				if dp[i-1][j] > dp[i][j-1] {
					dp[i][j] = dp[i-1][j]
				} else {
					dp[i][j] = dp[i][j-1]
				}
			}
		}
	}
	return dp[n][m]
}

// ── Расстояние Левенштейна ─────────────────────────────────────────
func EditDistance(a, b string) int {
	n, m := len(a), len(b)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, m+1)
		dp[i][0] = i
	}
	for j := 0; j <= m; j++ {
		dp[0][j] = j
	}
	min3 := func(x, y, z int) int {
		m := x
		if y < m {
			m = y
		}
		if z < m {
			m = z
		}
		return m
	}
	for i := 1; i <= n; i++ {
		for j := 1; j <= m; j++ {
			if a[i-1] == b[j-1] {
				dp[i][j] = dp[i-1][j-1]
			} else {
				dp[i][j] = 1 + min3(dp[i-1][j], dp[i][j-1], dp[i-1][j-1])
			}
		}
	}
	return dp[n][m]
}

// ── LIS за O(n log n) ──────────────────────────────────────────────
func LIS(a []int) int {
	tails := []int{}
	for _, x := range a {
		i := sort.SearchInts(tails, x)
		if i == len(tails) {
			tails = append(tails, x)
		} else {
			tails[i] = x
		}
	}
	return len(tails)
}

func main() {
	fmt.Println("Fib(20):", Fib(20))
	fmt.Println("CoinChange [1,3,4], 6:", CoinChange([]int{1, 3, 4}, 6))
	fmt.Println("Knapsack:", Knapsack01([]int{2, 3, 4, 5}, []int{3, 4, 5, 6}, 5))
	fmt.Println("LCS ABCBDAB / BDCABA:", LCS("ABCBDAB", "BDCABA"))
	fmt.Println("EditDistance kitten/sitting:", EditDistance("kitten", "sitting"))
	fmt.Println("LIS:", LIS([]int{10, 9, 2, 5, 3, 7, 101, 18}))
}
